### PR TITLE
Add an advertise-address flag

### PR DESCRIPTION
**Pull request by cjcullen**
This allows the address that the apiserver binds to (possibly 0.0.0.0) to be different than the address on which members of the cluster can reach the apiserver (possibly not a local interface).

### CODE REVIEW

File: pkg/master/master.go [hunk 1]
```diff
@@ -132,7 +132,9 @@ type Config struct {
 	// ExternalHost is the host name to use for external (public internet) facing URLs (e.g. Swagger)
 	ExternalHost string
 
-	// If nil, the first result from net.InterfaceAddrs will be used.
+	// PublicAddress is the IP address where members of the cluster (kubelet,
+	// kube-proxy, services, etc.) can reach the master.
+	// If nil or 0.0.0.0, the first result from net.InterfaceAddrs will be used.
```
Comment (thockin) [inline pkg/master/master.go]: It's not precisely the first interface.  More like "the host's default interface"

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -137,8 +138,13 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
...
 	fs.Var(&s.InsecureBindAddress, "address", "DEPRECATED: see --insecure-bind-address instead")
 	fs.Var(&s.BindAddress, "bind-address", ""+
-		"The IP address on which to serve the --read-only-port and --secure-port ports. This "+
-		"address must be reachable by the rest of the cluster. If blank, all interfaces will be used.")
+		"The IP address on which to serve the --read-only-port and --secure-port ports. The "+
+		"associated interface(s) must be reachable by the rest of the cluster, and by CLI/web "+
+		"clients. If blank, all interfaces will be used (0.0.0.0).")
+	fs.Var(&s.AdvertiseAddress, "advertise-address", ""+
+		"The IP address on which to advertise the apiserver to members of the cluster. This "+
+		"address must be reachable by the rest of the cluster. If blank, all interfaces will be "+
```
Comment (thockin) [inline cmd/kube-apiserver/app/server.go]: What does "all interfaces" mean here?  It should probably say and be "If blank, the --bind-address will be used.  If that is not specified, the host's default interface will be used."

Then implement that below.

File: pkg/master/master.go (hunk 1 above)
Comment (cjcullen) [inline pkg/master/master.go]: I went ahead and just killed the TODO from below. Now, PublicAddress must be set.

File: cmd/kube-apiserver/app/server.go (hunk 2 above)
Comment (cjcullen) [inline cmd/kube-apiserver/app/server.go]: Ahh. That means I forgot to update the second sentence of the description.

Fixed as recommended.

File: cluster/gce/configure-vm.sh [hunk 3]
```diff
@@ -488,6 +488,7 @@ project-id = ${PROJECT_ID}
 EOF
     cat <<EOF >>/etc/salt/minion.d/grains.conf
   cloud_config: /etc/gce.conf
+  advertise_address: '${KUBERNETES_MASTER_NAME}'
```
Comment (cjcullen) [inline cluster/gce/configure-vm.sh]: Right now, this is inside the "if (using alternate tokenSource)", which happens to be when we set KUBERNETES_MASTER_NAME to be an IP on GKE.

There's probably a better way to organize this, but this seemed expedient.

File: pkg/master/master.go (hunk 1 above)
Comment (cjcullen) [inline pkg/master/master.go]: Nevermind. Too many integration tests rely on the magic behavior.

**Comment by cjcullen** [comment]
e2e's aren't too happy with this right now. Can't tell if that is me or upstream.

File: pkg/master/master.go [hunk 4]
```diff
@@ -132,7 +132,9 @@ type Config struct {
 	// ExternalHost is the host name to use for external (public internet) facing URLs (e.g. Swagger)
 	ExternalHost string
 
-	// If nil, the first result from net.InterfaceAddrs will be used.
+	// PublicAddress is the IP address where members of the cluster (kubelet,
+	// kube-proxy, services, etc.) can reach the master.
+	// If nil or 0.0.0.0, the host's default interface will be used.
```
Comment (thockin) [inline pkg/master/master.go]: I kinda dislike that ChooseHostInterface() is getting called in multiple places now: the one you added in cmd/kube-apiserver/app/server.go and the existing one in pkg/master/master.go.  Is it safe to get rid of the one in master,go and just say that PublicAddress is required?  I wonder if that will break tests.

Comment (thockin) [inline pkg/master/master.go]: Oopps, I read further and saw you did this and then reverted.

Comment (thockin) [inline pkg/master/master.go]: Could we move the verifyAdvertiseAddress() logic into master.go setDefaults() ?  That way it's all in one core place.

Comment (cjcullen) [inline pkg/master/master.go]: Okay. I changed it back closer to what it was before. The apiserver will set advertise-address = bind-address if advertise-address was unset. Then master.go setDefaults() will do the util.ChooseHostInterface() dance as it does currently if advertise-address is still 0.0.0.0.

**Comment by thockin** [comment]
LGTM

**Comment by cjcullen** [comment]
Summarizing 4 Failures:

[Fail] PD [It] should schedule a pod w/ a readonly PD on two hosts, then remove both. 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:170

[Panic!] Density [BeforeEach] [Performance suite] should allow starting 3 pods per node 
/usr/src/go/src/runtime/panic.go:387

[Panic!] Density [BeforeEach] [Performance suite] should allow starting 30 pods per node 
/usr/src/go/src/runtime/panic.go:387

[Fail] PD [It] should schedule a pod w/ a RW PD, remove it, then schedule it on another host 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:105

Ran 45 of 51 Specs in 2759.910 seconds
FAIL! -- 41 Passed | 4 Failed | 1 Pending | 5 Skipped --- FAIL: TestE2E (2770.54s)
FAIL

This change shouldn't change behavior of e2e's at all (if I didn't mess up), but I'll investigate a bit.

**Comment by thockin** [comment]
status?

**Comment by cjcullen** [comment]
Haven't changed anything. I'll rebase and rerun E2Es to see if I can get a more confident answer.

**Comment by cjcullen** [comment]
Summarizing 5 Failures:

[Fail] DNS [It] should provide DNS for the cluster 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/dns.go:174

[Fail] DNS [It] should provide DNS for headless services 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/dns.go:335

[Fail] PD [It] should schedule a pod w/ a readonly PD on two hosts, then remove both. 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:170

[Fail] PD [It] should schedule a pod w/ a RW PD, remove it, then schedule it on another host 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:105

[Fail] Shell [It] tests that services.sh passes 
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/shell.go:50

Ran 52 of 60 Specs in 3744.696 seconds
FAIL! -- 47 Passed | 5 Failed | 1 Pending | 7 Skipped --- FAIL: TestE2E (3745.21s)

**Comment by cjcullen** [comment]
I'm fine with letting this PR simmer until we get 0.18.0 out and we have a better signal from e2e's

**Comment by thockin** [comment]
This still LGTM and I don't see how those flaky tests are your fault, I would rebase and push and go for merge

**Comment by erictune** [comment]
@cjcullen needs rebase

**Comment by cjcullen** [comment]
Rebased and added external-IP-getting call (KUBERNETES_MASTER_NAME is no longer getting stuck in the master's env).

PTA(quick)L

**Comment by thockin** [comment]
Still LGTM

**Pull request merged**